File: 95. Unique Binary Search Trees II.py
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        if n == 0:
            return []
        def generateTrees(self, mn: int, mx :int) -> List[Optional[TreeNode]]:
            if mn > mx:
                return [None]
            res = []

            for i in range(mn, mn + 1):
                for left in generateTrees(mn, i - 1):
                    for right in generateTrees(i + 1, mx):
                        res.append(TreeNode(i))
                        res[-1].left = left
                        res[-1].right = right
            
        return generateTrees(1, n)
        

File: 95. Unique Binary Search Trees II.go
type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}
func generateTrees(n int) []*TreeNode {
	if n == 0{return [] *TreeNode{nil}}
	fn = func(mn, mx int) []* TreeNode{
		if(mn > mx){
			return []*TreeNode{nil};
		}	
	
		res := []* TreeNode{};
	
		for i := mn; i <= mx; i++{
			leftTrees := fn(mn, i-1);
			rightTrees := fn(i+1, mx);
			for _, left := range leftTrees{
				for _, right := range rightTrees{
					root := &TreeNode{Val: i, Left: left, Right: right};
					res = append(res, root)
				}
			}
	
		}
	
		return res
	}
	return fn(1, n)
}

